File: game_client/Assets/Scripts/ResScripts/WCGResCommon/Pair/WCGPairInt.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using Random = UnityEngine.Random;

[Serializable]
public class WCGPairInt
{
    [SerializeField]
    private WCGPair<int, int> _m_pair = new WCGPair<int, int>(0, 0);

    public WCGPairInt()
    {

    }

    public WCGPairInt(int first, int second)
    {
        _m_pair.first = first;
        _m_pair.second = second;
    }

    public int first()
    {
        return _m_pair.first;
    }

    public int second()
    {
        return _m_pair.second;
    }

    public void setSecond(int _second)
    {
        _m_pair.second = _second;
    }

    public void setFirst(int _first)
    {
        _m_pair.first = _first;
    }

    public void ParseFromString(string _str)
    {
        if (string.IsNullOrEmpty(_str))
            return;

        try
        {
            string[] strs = _str.Split(new string[] { ":" }, StringSplitOptions.RemoveEmptyEntries);
            if (strs.Length >= 1)
            {
                if (!string.IsNullOrEmpty(strs[0]))
                    _m_pair.first = int.Parse(strs[0]);
            }
            if (strs.Length >= 2)
            {
                if (!string.IsNullOrEmpty(strs[1]))
                    _m_pair.second = int.Parse(strs[1]);
            }
        }
        catch (Exception e)
        {
            Debug.LogError($"WCGPairLong  parse failed,{_str}\n{e}");
        }
    }
    public static WCGPairInt readFromStr(string _str)
    {
        //拆分字符串后进行读取
        string[] strs = _str.Split(new string[] { ":" }, StringSplitOptions.RemoveEmptyEntries);

        if (strs.Length < 1)
        {
            UnityEngine.Debug.LogWarning("没有配置 对象名!");
            return null;
        }

        //需要支持物品类型没有配置的情况
        WCGPairInt ret = new WCGPairInt();
        try
        {
            if (strs.Length > 0)
                ret._m_pair.first = int.Parse(strs[0]);

            if (strs.Length > 1)
                ret._m_pair.second = int.Parse(strs[1]);
        }
        catch (Exception e)
        {
            Debug.LogError($"解析WCGPairInt时发生错误：{_str}\n{e}");
            return null;
        }

        return ret;
    }

    public bool isDefault()
    {
        return _m_pair.first == 0 && _m_pair.second == 0;
    }

    /// <summary>
    /// 读取队列
    /// </summary>
    /// <param name="_str"></param>
    /// <returns></returns>
    public static List<WCGPairInt> readList(string _str)
    {
        List<WCGPairInt> list = new List<WCGPairInt>();
        if (null == _str || _str.Length <= 0)
            return list;

        string[] strs = _str.Split(new string[] { "|", ";" }, StringSplitOptions.RemoveEmptyEntries);
        for (int i = 0; i < strs.Length; i++)
        {
            WCGPairInt newItem = WCGPairInt.readFromStr(strs[i]);
            if (null == newItem)
                continue;

            list.Add(newItem);
        }
        return list;
    }


    /// <summary>
    /// 为自动导出写的
    /// </summary>
    public static List<WCGPairInt> MakeListFromString(string _str)
    {
        return readList(_str);
    }

    /// <summary>
    /// 为自动导出写的
    /// </summary>
    public static WCGPairInt[] MakeArrayFromString(string _str)
    {
        return readList(_str).ToArray();
    }

    public override string ToString()
    {
        return _m_pair.ToString();
    }

    public int rand()
    {
        int delta = _m_pair.second - _m_pair.first;
        if (delta > 0)
        {
            return _m_pair.first + Random.Range(0, delta + 1);
        }
        else
        {
            return _m_pair.first;
        }
    }
    
    /**
 * 把传入值限制在范围内
 */
    public int limit(double _value)
    {
        if (_value < _m_pair.first)
            return _m_pair.first;
        if (_value > _m_pair.second)
            return _m_pair.second;
        return (int) _value;
    }
}

File: game_server/Common/src/NPCommon/Util/Pair/WCGPairInt.java
package NPCommon.Util.Pair;


import NPCommon.Log.CommLog;
import NPCommon.RefData._IParseFromStringable;
import NPCommon.Util.CommonFunc;

/**
 * 整数对
 * 用于表示范围（无限）时 只能用于general表中 单字段判断
 */
public class WCGPairInt implements _IParseFromStringable
{
    private WCGPair<Integer, Integer> _m_pair = new WCGPair<Integer, Integer>(0, 0);

    public WCGPairInt()
    {

    }

    public WCGPairInt(int first, int second)
    {
        _m_pair.first = first;
        _m_pair.second = second;
    }

    public int first()
    {
        return _m_pair.first;
    }

    public int second()
    {
        return _m_pair.second;
    }

    public void setSecond(int _second)
    {
        _m_pair.second = _second;
    }

    public void addSecond(int _second)
    {
        _m_pair.second += _second;
    }

    public void setFirst(int _first)
    {
        _m_pair.first = _first;
    }

    public void addFirst(int _first)
    {
        _m_pair.first += _first;
    }
    
    public void add(WCGPairInt _pair)
    {
    	_m_pair.first += _pair.first();
    	_m_pair.second += _pair.second();
    }

    @Override
    public boolean parseFromString(String sValue)
    {
        return parseFromString(sValue, ':');
    }

    public boolean parseFromString(String sValue, char token)
    {
        if (null == sValue || sValue.trim().isEmpty())
            return true;
        try
        {
            String[] strs = CommonFunc.charSplit(sValue, token, 2);
            if (strs.length >= 1)
            {
                if (!strs[0].isEmpty())
                    _m_pair.first = Integer.parseInt(strs[0].trim());
            }
            if (strs.length >= 2)
            {
                if (!strs[1].isEmpty())
                    _m_pair.second = Integer.parseInt(strs[1].trim());
            }
        } catch (Exception e)
        {
            CommLog.error("WCGPairLong  parse failed,str =" + sValue, e);
            return false;
        }

        return true;
    }

    public boolean isDefault()
    {
        return _m_pair.first == 0 && _m_pair.second == 0;
    }

    public static WCGPairInt fromString(String sValue)
    {
        return fromString(sValue, ':');
    }

    public static WCGPairInt fromString(String sValue, char token)
    {
        WCGPairInt pair = new WCGPairInt();
        if (!pair.parseFromString(sValue, token))
        {
            return null;
        }
        return pair;
    }

    @Override
    public String toString()
    {
        return _m_pair.toString();
    }

    public int rand()
    {
        int delta = _m_pair.second - _m_pair.first;
        if (delta > 0)
        {
            return _m_pair.first + CommonFunc.randomInt(delta);
        } else
        {
            return _m_pair.first;
        }
    }

    /**
     * 是否在范围内 支持不限制范围，-1表示不限制
     * @param _value
     * @return
     */
    public boolean inRange(int _value)
    {
        if (_m_pair.first == -1 && _m_pair.second == -1)
            return true;

        if (_m_pair.first != -1 && _value < _m_pair.first)
            return false;

        if (_m_pair.second != -1 && _value > _m_pair.second)
            return false;

        return true;
    }

    /**
     * 把传入值限制在范围内 支持不限制范围，-1表示不限制
     */
    public long limit(long _value)
    {
        if (_m_pair.first == -1 && _m_pair.second == -1)
            return _value;

        if (_m_pair.first != -1 && _value < _m_pair.first)
            return _m_pair.first;

        if (_m_pair.second != -1 && _value > _m_pair.second)
            return _m_pair.second;

        return _value;
    }
    
    public void clear()
    {
    	_m_pair.first = 0;
    	_m_pair.second = 0;
    }
}